File: wasm/verify/permission_test.mjs
// The browser half of Go's permission package, against the minimal DOM.
//
// This is the host where the two commands are genuinely different operations
// and where one of them mostly cannot be honoured, so almost everything worth
// pinning here is a browser fact that no Go test can reach:
//
//   check     navigator.permissions.query — a real read, no UI, answering in
//             exactly the words Go's Status carries
//   request   no such API. A page obtains camera, microphone or location by
//             *calling the feature*, and the prompt is a side effect of that
//             call — so a request is a getUserMedia whose tracks are stopped
//             the instant it resolves, or a single getCurrentPosition
//
// The interesting failures are all in the seams: a descriptor Firefox has
// never heard of makes query() *throw* rather than resolve, a getUserMedia
// rejection says "denied" and "there is no camera" through the same channel,
// and a granted media request has genuinely opened a capture device that has
// to be closed again.

import test from "node:test";
import assert from "node:assert/strict";

import { loadRuntime } from "./load.mjs";

// A runtime with a place for permission answers to land, plus whatever
// navigator APIs the case is modelling. Returns the events Go would have
// seen, decoded.
function harness(navigator = {}) {
    const rt = loadRuntime();
    const answers = [];
    rt.window.GrMobWASM = {
        HostEvent: (name, json) => {
            if (name === "permission") answers.push(JSON.parse(json));
        },
    };
    Object.assign(rt.sandbox.navigator, navigator);
    return {
        rt,
        answers,
        check: (kind) => rt.GrMob.permission.handle({ command: "check", kind }),
        request: (kind) => rt.GrMob.permission.handle({ command: "request", kind }),
    };
}

// The runtime answers from promise callbacks, so every assertion has to come
// after the microtask queue has drained. Two turns rather than one: the media
// path chains a .then onto the getUserMedia promise, so its report lands one
// tick behind the call.
const settle = async () => { await null; await null; await null; };

// --- check ------------------------------------------------------------------

test("a check answers with the Permissions API's own word", async () => {
    // The whole argument for Go spelling its Status values the way the W3C
    // Permissions API does: this host needs no mapping table, and the state
    // is forwarded unchanged.
    for (const state of ["granted", "denied", "prompt"]) {
        const h = harness({
            permissions: { query: () => Promise.resolve({ state }) },
        });
        h.check("camera");
        await settle();
        assert.deepEqual(h.answers, [{ kind: "camera", status: state }]);
    }
});

test("a check uses the descriptor name, not the Go kind", async () => {
    // "location" is Go's word and "geolocation" is the browser's; a query for
    // the wrong one throws, which the catch below would turn into a plausible
    // "unavailable" rather than an error anybody would see.
    let asked;
    const h = harness({
        permissions: {
            query: (d) => { asked = d.name; return Promise.resolve({ state: "granted" }); },
        },
    });
    h.check("location");
    await settle();
    assert.equal(asked, "geolocation");
    assert.deepEqual(h.answers, [{ kind: "location", status: "granted" }]);
});

test("a descriptor this browser has never heard of is unavailable", async () => {
    // query() rejects — with a TypeError, not a resolved state — for a name it
    // does not know, and browsers disagree about which names those are:
    // Firefox has no "camera" descriptor at all. So this is the common path on
    // some browsers rather than an edge case, and it must not surface as a
    // denial: a denial sends the user to a settings page that has no switch
    // on it.
    const h = harness({
        permissions: { query: () => Promise.reject(new TypeError("bad descriptor")) },
    });
    h.check("camera");
    await settle();
    assert.deepEqual(h.answers, [{ kind: "camera", status: "unavailable" }]);
});

test("a page with no Permissions API at all answers unavailable", async () => {
    const h = harness(); // the default navigator: nothing installed
    h.check("microphone");
    await settle();
    assert.deepEqual(h.answers, [{ kind: "microphone", status: "unavailable" }]);
});

test("storage has no descriptor and says so without asking", async () => {
    // A page reaches files through an <input> or the file-system access API,
    // both of which are a gesture rather than a permission. There is nothing
    // to query and nothing to request, so the honest answer is that this
    // platform cannot grant it — and it must be *sent*, or a screen waiting on
    // it waits forever.
    let queried = false;
    const h = harness({
        permissions: { query: () => { queried = true; return Promise.resolve({ state: "granted" }); } },
    });
    h.check("storage");
    await settle();
    assert.equal(queried, false, "storage must not reach the Permissions API");
    assert.deepEqual(h.answers, [{ kind: "storage", status: "unavailable" }]);
});

// --- request ----------------------------------------------------------------

test("a media request prompts through getUserMedia and closes the device again", async () => {
    // The prompt is the point and the stream is not: a resolved getUserMedia
    // means a live capture device, and leaving it running would keep the
    // browser's recording indicator lit for a page that only wanted an answer.
    let stopped = 0;
    let constraints;
    const h = harness({
        mediaDevices: {
            getUserMedia: (c) => {
                // Serialized at the boundary, as load.mjs does for dispatches
                // and for the same reason: an object built inside the vm
                // context carries that realm's Object.prototype, so a strict
                // deep-equal against a literal here fails on prototype
                // identity alone.
                constraints = JSON.parse(JSON.stringify(c));
                return Promise.resolve({ getTracks: () => [{ stop: () => { stopped++; } }] });
            },
        },
    });
    h.request("camera");
    await settle();

    assert.deepEqual(constraints, { video: true });
    assert.equal(stopped, 1, "the track was left running after the prompt");
    assert.deepEqual(h.answers, [{ kind: "camera", status: "granted" }]);
});

test("the microphone asks for audio, not video", async () => {
    let constraints;
    const h = harness({
        mediaDevices: {
            getUserMedia: (c) => {
                constraints = JSON.parse(JSON.stringify(c));
                return Promise.resolve({ getTracks: () => [] });
            },
        },
    });
    h.request("microphone");
    await settle();
    assert.deepEqual(constraints, { audio: true });
});

test("a refused prompt and an absent device are told apart", async () => {
    // Both arrive as a rejected getUserMedia and they need different words,
    // because only one of them has a fix in the browser's settings. That
    // distinction is the whole reason Go's Status has Unavailable beside
    // Denied.
    for (const [name, status] of [
        ["NotAllowedError", "denied"],
        ["NotFoundError", "unavailable"],
        ["OverconstrainedError", "unavailable"],
        ["NotReadableError", "unavailable"],
        // Anything unrecognised falls to denied: an unknown failure of a
        // prompt is much more likely to be a refusal than a missing camera,
        // and denied is the answer whose remedy is harmless to offer.
        ["SomeFutureError", "denied"],
    ]) {
        const h = harness({
            mediaDevices: {
                getUserMedia: () => Promise.reject(Object.assign(new Error("no"), { name })),
            },
        });
        h.request("camera");
        await settle();
        assert.deepEqual(h.answers, [{ kind: "camera", status }], name);
    }
});

test("a location request prompts through getCurrentPosition", async () => {
    const h = harness({
        geolocation: { getCurrentPosition: (ok) => ok({ coords: {} }) },
    });
    h.request("location");
    await settle();
    assert.deepEqual(h.answers, [{ kind: "location", status: "granted" }]);
});

test("only PERMISSION_DENIED is a denial; a failed fix re-checks", async () => {
    // POSITION_UNAVAILABLE and TIMEOUT are failures of the *fix* and not of
    // the permission — the page may be perfectly authorised and simply
    // indoors — so reporting them as denied would tell a user to change a
    // setting that is already right.
    const denied = harness({
        geolocation: { getCurrentPosition: (_, fail) => fail({ code: 1 }) },
    });
    denied.request("location");
    await settle();
    assert.deepEqual(denied.answers, [{ kind: "location", status: "denied" }]);

    const indoors = harness({
        geolocation: { getCurrentPosition: (_, fail) => fail({ code: 2 }) },
        permissions: { query: () => Promise.resolve({ state: "granted" }) },
    });
    indoors.request("location");
    await settle();
    assert.deepEqual(indoors.answers, [{ kind: "location", status: "granted" }],
        "a position failure must fall back to reading the permission, not assume a refusal");
});

test("a page with no media or geolocation provider answers unavailable", async () => {
    const h = harness();
    h.request("camera");
    h.request("location");
    await settle();
    assert.deepEqual(h.answers, [
        { kind: "camera", status: "unavailable" },
        { kind: "location", status: "unavailable" },
    ]);
});

// --- the dispatcher ---------------------------------------------------------

test("an unknown kind or command is dropped, not answered", async () => {
    // Every host's contract for an unknown event. Answering would be worse
    // than silence here: Go validates the kind on arrival and would drop the
    // reply anyway, and a host that invented a status for a permission it does
    // not know is a host that will keep inventing one after Go learns about it.
    const h = harness({
        permissions: { query: () => Promise.resolve({ state: "granted" }) },
    });
    h.rt.GrMob.permission.handle({ command: "check", kind: "bluetooth" });
    h.rt.GrMob.permission.handle({ command: "revoke", kind: "camera" });
    h.rt.GrMob.permission.handle({});
    await settle();
    assert.deepEqual(h.answers, []);
});

test("an answer with no host channel attached is dropped rather than thrown", async () => {
    // A page that has not finished wiring GrMobWASM, which is a real window
    // during startup. Nothing here may throw into the runtime's own promise
    // chain, where the rejection would be unhandled and invisible.
    const rt = loadRuntime();
    Object.assign(rt.sandbox.navigator, {
        permissions: { query: () => Promise.resolve({ state: "granted" }) },
    });
    rt.GrMob.permission.handle({ command: "check", kind: "camera" });
    await settle();
});

File: wasm/verify/permission_test.go
package main

import (
	"strings"
	"testing"

	"github.com/rohanthewiz/grmob/permission"
)

// The browser half of Go's permission package, pinned from Go.
//
// permission_test.mjs proves this module *behaves* — that a query becomes the
// right descriptor, that a rejected getUserMedia is told apart from an absent
// camera, that a stream is closed again — but it runs only under run.sh, which
// needs Node and which a human has to remember. This is the half
// `go test ./...` reaches, and what it holds is the contract rather than the
// implementation: the two Go vocabularies, and the wire names on both sides.
//
// Every drift here is silent in exactly the same way. A kind the host has no
// entry for is dropped, and the screen sits on permission.Unknown drawing the
// spinner that state is for; a status the host spells wrongly is dropped by
// permission.reportable, and the record never moves.

// Every permission.Permission needs a descriptor-table entry, including the
// one whose entry is deliberately null.
//
// The null entry is the case worth stating: `storage` has no browser
// permission at all, and the difference between "absent from the table" and
// "present and null" is the difference between a screen waiting forever and a
// screen told this platform cannot do it. The host answers the second and
// drops the first, so the entry has to be there.
func TestTheBrowserHostCoversEveryPermission(t *testing.T) {
	src := runtimeSource(t)
	for _, p := range permission.Permissions() {
		if !strings.Contains(src, string(p)+": ") {
			t.Errorf("grmob-runtime.js: DESCRIPTORS has no entry for permission.Permission "+
				"%q — the kind is dropped, and a dropped kind is indistinguishable from "+
				"one the user has not answered", p)
		}
	}
	if !strings.Contains(src, "storage: null") {
		t.Error("grmob-runtime.js: storage must map to null rather than be absent — a null " +
			"entry is answered \"unavailable\", an absent one is silently dropped")
	}
}

// Every status the host can report must be one permission.Statuses() declares.
//
// The three the Permissions API produces are forwarded verbatim rather than
// mapped, which is the whole payoff of Go spelling Status the way the browser
// does — so what is checked is the places the host writes a word itself, and
// that the fourth, which the browser has no word for, is spelled the way Go
// does.
func TestTheBrowserHostReportsOnlyDeclaredStatuses(t *testing.T) {
	src := runtimeSource(t)
	for _, want := range []struct{ expr, why string }{
		{`report(kind, "granted")`,
			"a resolved getUserMedia. The query path forwards the browser's own state " +
				"untouched, so this is one of the few places the host writes a word itself"},
		{`? "unavailable" : "denied"`,
			"the split that makes a missing camera a different answer from a refused " +
				"prompt — only one of the two has a fix in the browser's settings"},
		{`report("location", "denied")`,
			"PERMISSION_DENIED, which is the only geolocation error that means a refusal"},
		{`report(kind, "unavailable")`,
			"a page with no media provider at all"},
		{`Promise.resolve("unavailable")`,
			"a descriptor this browser does not have. query() throws for an unknown name " +
				"rather than resolving, and Firefox has no camera descriptor, so this is " +
				"a common path and not an edge case"},
	} {
		if !strings.Contains(src, want.expr) {
			t.Errorf("grmob-runtime.js: %q not found — %s", want.expr, want.why)
		}
	}

	if strings.Contains(src, `report(kind, "prompt")`) {
		t.Error("grmob-runtime.js: the host writes \"prompt\" itself somewhere — that word " +
			"should only ever reach Go as the Permissions API's own answer, forwarded")
	}
}

// The wire, on both sides of it. The names are string literals in four places
// (Go, this runtime, and two native shells) and nothing but a check like this
// connects them.
func TestTheBrowserHostSpeaksThePermissionWire(t *testing.T) {
	src := runtimeSource(t)
	for _, want := range []struct{ expr, why string }{
		{`if (name === "permission") {`,
			`the system-event arm. Without it permission.Check is dropped at the page ` +
				`boundary, exactly as it would be on a native shell with no case`},
		{`host.HostEvent("permission", JSON.stringify({ kind, status }))`,
			"the answer's shape — both keys are required, and Go drops a payload missing " +
				"either rather than guessing at it"},
		{`case "check": check(cmd.kind); break;`,
			"the check command, which must never prompt: it is what a screen calls on mount"},
		{`case "request": request(cmd.kind); break;`,
			"the request command, which is the only one that puts a dialog up"},
	} {
		if !strings.Contains(src, want.expr) {
			t.Errorf("grmob-runtime.js: %q not found — %s", want.expr, want.why)
		}
	}
}

// A request must actually prompt, which on this platform means calling the
// feature rather than asking about it.
//
// There is no requestPermission API in a browser: camera, microphone and
// location are obtained by using them, and the prompt is a side effect. A host
// that quietly answered a request with a query would pass every shape check
// above and leave a button that does nothing at all.
func TestABrowserRequestCallsTheFeatureRatherThanQueryingIt(t *testing.T) {
	src := runtimeSource(t)
	for _, want := range []struct{ expr, why string }{
		{"md.getUserMedia(constraints)",
			"the camera and microphone prompt"},
		{"stream.getTracks().forEach((t) => t.stop())",
			"closing the device again. A resolved getUserMedia is a live capture device, " +
				"and a page that only wanted an answer must not leave the browser's " +
				"recording indicator lit"},
		{"navigator.geolocation.getCurrentPosition(",
			"the location prompt"},
	} {
		if !strings.Contains(src, want.expr) {
			t.Errorf("grmob-runtime.js: %q not found — %s", want.expr, want.why)
		}
	}
}
